flink-runtime/src/main/java/org/apache/flink/runtime/metrics/ReporterSetup.java

```diff
@@ -210,17 +213,31 @@ private static ReporterSetup createReporterSetup(String reporterName, MetricConf
 		return namedOrderedReporters;
 	}
 
-	private static Map<String, MetricReporterFactory> loadReporterFactories() {
-		final ServiceLoader<MetricReporterFactory> serviceLoader = ServiceLoader.load(MetricReporterFactory.class);
-
+	private static Map<String, MetricReporterFactory> loadAvailableReporterFactories(PluginManager pluginManager) {
 		final Map<String, MetricReporterFactory> reporterFactories = new HashMap<>(2);
-		final Iterator<MetricReporterFactory> factoryIterator = serviceLoader.iterator();
+		final Iterator<MetricReporterFactory> factoryIterator = getAllReporterFactories(pluginManager);
+		LOG.info("Prepare reporter factories (from both SPIs and Plugins):");
 		// do not use streams or for-each loops here because they do not allow catching individual ServiceConfigurationErrors
 		// such an error might be caused if the META-INF/services contains an entry to a non-existing factory class
 		while (factoryIterator.hasNext()) {
 			try {
 				MetricReporterFactory factory = factoryIterator.next();
-				reporterFactories.put(factory.getClass().getName(), factory);
+				String factoryClassName = factory.getClass().getName();
+				MetricReporterFactory existingFactory = reporterFactories.get(factoryClassName);
+				if (existingFactory == null){
+					reporterFactories.put(factoryClassName, factory);
+					LOG.info("Found reporter factory {} at {} ",
+						factoryClassName,
+						new File(factory.getClass().getProtectionDomain().getCodeSource().getLocation().toURI()).getCanonicalPath());
+				} else {
+					//TODO: use path information below, when Plugin Classloader stops always prioritizing factories from /lib
+//					String jarPath1 = new File(existingFactory.getClass().getProtectionDomain().getCodeSource().getLocation()
+//						.toURI()).getCanonicalPath();
+//					String jarPath2 = new File(factory.getClass().getProtectionDomain().getCodeSource().getLocation()
+//						.toURI()).getCanonicalPath();
+//					LOG.warn("Multiple implementations of the same reporter were found: \n {} and \n{}", jarPath1, jarPath2);
+					LOG.warn("Multiple implementations of the same reporter were found in 'lib' and/or 'plugins' directories for {}. It is recommended to remove redundant reporter JARs to resolve used versions' ambiguity.", factoryClassName);
```

Comment:
`/` means `or`. `and/or` is `and or or`, which can be simplified to `or`. (`or` is not `xor` in English)

Reply:
It's frequently used in our code and documentation; I don't think we gain anything by watching out for things like this.
Personally I find it less ambiguous than just `or`.

Reply:
I wouldn't care if it's just in a comment or internal exception. But if it's user facing, I'd strongly recommend to proof-read everything and fix it.

Reply:
https://en.wikipedia.org/wiki/And/or 

`It is used as an inclusive "or" (as in logic and mathematics), while an "or" in spoken language might be inclusive or exclusive.`

Seems to me like something that reduces ambiguity, as @zentol said.

Reply:
No hard feeling. You can safely ignore that comment. Just wanted to point it out and closing with a cite of your link 

> It has been strongly criticized as both ugly in style, and ambiguous in legal documents .

Same in academic writing.

Reply:
I thing the problem is that we have too many variants to describe with a single coordinating conjunction like "or" (plus "or" can unfortunately be both inclusive and exclusive). Cases:
(xx) _ ()
() _ (xx)
(x) _ (x)
(xx) _ (xx)

Alternatives like "x or y or both" also do not work, because in this particular case they can be easily be interpreted as if the "(x) _ (x)" case is not an issue (because of "Multiple" in the beginning). Seems to me like the best case to indicate this ambiguity is is to use "and/or" so that people will be aware for what kinds of problem potentially to look for.

Reply:
I think we should stop discussing this here; this is relevant for the entirety of the documentation, and we could get much more experienced people involved if we target that instead.

Reply:
I leave it as "and/or" here, and then we comb through the docs and sources to address it separately.